scheduler/propertyset.go

```diff
@@ -0,0 +1,265 @@
+			continue
+		}
+
+		p.existingValues[nProperty] = struct{}{}
+	}
+}
+
+// PopulateProposed populates the proposed values and recomputes any cleared
+// value. It should be called whenever the plan is updated to ensure correct
+// results when checking an option.
+func (p *propertySet) PopulateProposed() {
+
+	// Reset the proposed properties
+	p.proposedValues = make(map[string]struct{})
+	p.clearedValues = make(map[string]struct{})
+
+	// Gather the set of proposed stops.
+	var stopping []*structs.Allocation
+	for _, updates := range p.ctx.Plan().NodeUpdate {
+		stopping = append(stopping, updates...)
+	}
+	stopping = p.filterAllocs(stopping)
+
+	// Gather the proposed allocations
+	var proposed []*structs.Allocation
+	for _, pallocs := range p.ctx.Plan().NodeAllocation {
+		proposed = append(proposed, pallocs...)
+	}
+	proposed = p.filterAllocs(proposed)
+
+	// Get the used nodes
+	both := make([]*structs.Allocation, 0, len(stopping)+len(proposed))
+	both = append(both, stopping...)
+	both = append(both, proposed...)
+	nodes, err := p.buildNodeMap(both)
+	if err != nil {
+		p.errorBuilding = err
+		p.ctx.Logger().Printf("[ERR] scheduler.dynamic-constraint: %v", err)
+		return
+	}
+
+	// Populate the cleared values
+	for _, alloc := range stopping {
+		nProperty, ok := p.getProperty(nodes[alloc.NodeID], p.constraint.LTarget)
+		if !ok {
+			continue
+		}
+
+		p.clearedValues[nProperty] = struct{}{}
+	}
+
+	// Populate the proposed values
+	for _, alloc := range proposed {
+		nProperty, ok := p.getProperty(nodes[alloc.NodeID], p.constraint.LTarget)
+		if !ok {
+			continue
+		}
+
+		p.proposedValues[nProperty] = struct{}{}
+
+		// If it was cleared, it is now being used
+		delete(p.clearedValues, nProperty)
+	}
+}
+
+// SatisfiesDistinctProperties checks if the option satisfies the
+// distinct_property constraints given the existing placements and proposed
+// placements. If the option does not satisfy the constraints an explanation is
+// given.
+func (p *propertySet) SatisfiesDistinctProperties(option *structs.Node, tg string) (bool, string) {
+	// Check if there was an error building
+	if p.errorBuilding != nil {
+		return false, p.errorBuilding.Error()
+	}
+
+	// Get the nodes property value
+	nValue, ok := p.getProperty(option, p.constraint.LTarget)
+	if !ok {
+		return false, fmt.Sprintf("missing property %q", p.constraint.LTarget)
+	}
+
+	// both is used to iterate over both the proposed and existing used
+	// properties
+	bothAll := []map[string]struct{}{p.existingValues, p.proposedValues}
+
+	// Check if the nodes value has already been used.
+	for _, usedProperties := range bothAll {
+		// Check if the nodes value has been used
+		_, used := usedProperties[nValue]
+		if !used {
+			continue
+		}
+
+		// Check if the value has been cleared from a proposed stop
+		if _, cleared := p.clearedValues[nValue]; cleared {
+			continue
+		}
+
+		return false, fmt.Sprintf("distinct_property: %s=%s already used", p.constraint.LTarget, nValue)
```

Comment:
I wonder if including the value which may be distinct could case an explosion of annotations for `AllocMetrics` e.g. if I have high cardinality. Might be not worth considering at this point.